hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -357,11 +357,12 @@ public final boolean complete(Object value) {
         for (; ; ) {
             final Object oldState = state;
             if (isDone(oldState)) {
-                // it can be that this invocation future already received an answer, e.g. when an invocation already received a
-                // response, but before it cleans up itself, it receives a HazelcastInstanceNotActiveException.
-                logger.warning("The Future.complete(Object value) can only be called once. Request: " + invocationToString()
-                        + ", current response: " + getState() + ", new response: " + value);
-
+                if (oldState != value) {
+                    // it can be that this future already completed, e.g. when an invocation already
+                    // received a response, but before it cleans up itself, it receives a HazelcastInstanceNotActiveException.
+                    logger.warning("Future.complete(Object value) can only be called once. Request: " + invocationToString()
+                            + ", current value: " + getState() + ", offered value: " + value);
```

Comment:
The volatile load is really not worth to mention.

1: it is cheap
2: this branch will only happen in very very rare cases
3: the performance of this branch will be totally dominated by the logging; not by the volatile read.

But I fully agree that it is a readability improvement not to use getState, but use the oldState variable.